ipapython/ipautil.py

```diff
@@ -282,6 +282,37 @@ def write_tmp_file(txt):
 
     return fd
 
+
+@contextmanager
+def concurrent_open(path, mode='r'):
+    """Ensure that complete file is read/written
+
+    Works only for r, rb, w, w+, wb, wb+ modes. Default is r.
+
+    In read mode behaves the same as build-in open. In write mode all data are
+    written to the temporary file first and then moved to target path.
+    This approach ensures that reading will be performed on complete file and
+    writting will result in complete or none file written at all.
+    """
+    if mode in ('w', 'w+', 'wb', 'wb+'):
+        with tempfile.NamedTemporaryFile(
+                mode=mode, prefix=path, delete=False) as temp_file:
```

Comment:
prefix is not enough. You have to include the ```dir``` arg to ensure that the file is created in the same directory as the target dir. Otherwise both files may end up on two diferent file systems and os.rename() is no longer atomic.

https://docs.python.org/3/library/tempfile.html#tempfile.mkstemp

You also can't just yield the temp file or its file attribute, because ```with concurrent_open(...)``` will call ```TemporaryFile.__exit__```, which closes the fd and breaks ```fsync()```.

```
class RenameContext(object):
    def __init__(self, temp_file, destpath):
        self.temp_file = temp_file
        self.destpath = destpath

    def __enter__(self):
        return self.temp_file.file  # raw file

    def __exit__(self, exc, value, tb):
        f = self.temp_file
        f.flush()
        os.fsync(f.fileno())
        f.close()
        try:
            os.rename(f.name, self.destpath)
        except OSError:
            os.unlink(f.name)  # only Windows developers call this remove :)
            raise
```

And then

```
filename = os.path.abspath(filename)
directory, prefix = os.path.splitpath(filename)
with tempfile.NamedTemporaryFile(mode=mode, dir=directory, prefix=prefix, delete=False) as f:
    return RenameContext(f, filename)
```